src/prog_models/models/uav_model/uav_model.py

```diff
@@ -0,0 +1,556 @@
+        psi = x['psi']
+        vx_a = x['vx']
+        vy_a = x['vy']
+        vz_a = x['vz']
+        p = x['p']
+        q = x['q']
+        r = x['r']
+
+        # Pre-compute Trigonometric values
+        # --------------------------------
+        sin_phi   = np.sin(phi)
+        cos_phi   = np.cos(phi)
+        sin_theta = np.sin(theta)
+        cos_theta = np.cos(theta)
+        tan_theta = np.tan(theta)
+        sin_psi   = np.sin(psi)
+        cos_psi   = np.cos(psi)
+        
+        # Compute drag forces
+        # -------------------
+        v_earth = np.array([vx_a, vy_a, vz_a]).reshape((-1,))
+        v_body = np.dot(geom.rot_eart2body_fast(sin_phi, cos_phi, sin_theta, cos_theta, sin_psi, cos_psi), v_earth)
+        fb_drag = self.vehicle_model.aero['drag'](v_body) 
+        fe_drag = np.dot(geom.rot_body2earth_fast(sin_phi, cos_phi, sin_theta, cos_theta, sin_psi, cos_psi), fb_drag)
+        fe_drag[-1] = np.sign(v_earth[-1]) * np.abs(fe_drag[-1])
+
+        # Update state vector
+        # -------------------
+        dxdt     = np.zeros((len(x),))
+        
+        dxdt[0] = vx_a
+        dxdt[1] = vy_a
+        dxdt[2] = vz_a
+        
+        dxdt[3]  = p + q * sin_phi * tan_theta + r * cos_phi * tan_theta
+        dxdt[4]  = q * cos_phi - r * sin_phi
+        dxdt[5]  = q * sin_phi / cos_theta + r * cos_phi / cos_theta
+        
+        dxdt[6]  = (sin_theta * cos_psi * cos_phi + sin_phi * sin_psi) * T / m - 1.0/m * fe_drag[0]
+        dxdt[7]  = (sin_theta * sin_psi * cos_phi - sin_phi * cos_psi) * T / m - 1.0/m * fe_drag[1]
+        dxdt[8]  = - self.parameters['gravity'] + cos_phi * cos_theta  * T / m - 1.0/m * fe_drag[2]
+
+        dxdt[9]  = (Iyy - Izz) / Ixx * q * r + tp * self.vehicle_model.geom['arm_length'] / Ixx
+        dxdt[10] = (Izz - Ixx) / Iyy * p * r + tq * self.vehicle_model.geom['arm_length'] / Iyy
+        dxdt[11] = (Ixx - Iyy) / Izz * p * q + tr *        1                / Izz
+        dxdt[12] = 1 
+
+        # Set vehicle state:
+        state_temp = np.array([x[iter] for iter in x.keys()])
+        self.vehicle_model.set_state(state=state_temp + dxdt*self.parameters['dt'])
+
+        return self.StateContainer(np.array([
+            np.atleast_1d(dxdt[0]),
+            np.atleast_1d(dxdt[1]),
+            np.atleast_1d(dxdt[2]),
+            np.atleast_1d(dxdt[3]),
+            np.atleast_1d(dxdt[4]),
+            np.atleast_1d(dxdt[5]),
+            np.atleast_1d(dxdt[6]),
+            np.atleast_1d(dxdt[7]),
+            np.atleast_1d(dxdt[8]),
+            np.atleast_1d(dxdt[9]),
+            np.atleast_1d(dxdt[10]),
+            np.atleast_1d(dxdt[11]),
+            np.atleast_1d(dxdt[12])
+        ]))
+    
+    def event_state(self, x : dict) -> dict:
+        # Extract next waypoint information 
+        num_wypts = len(self.parameters['waypoints']['waypoints_time']) - 1 # Don't include initial waypoint
+        index_next = self.parameters['waypoints']['next_waypoint']
```

Comment:
The model itself should be stateless. Some algorithms might require that you run multiple states through the model (e.g., when building a surrogate model). This will break down if you save state in the parameters. In this case, the next_waypoint is part of the state of the system. 

next_waypoint should be part of the state vector and updated in next_state, or it should be calculated each time. 

Since this is a bigger change, I'm ok with leaving it as is (with a note about this limitation) if you add this to the github issue for the next release.

Reply:
@teubert I'd love to talk with you more about this. For now, I've added it to the future work issue but hopefully we can get it done before the PR is finalized